### PR TITLE
[SPARK-8772][SQL] Implement implicit type cast for expressions that define input types.

### PR DESCRIPTION
**Pull request by rxin**

### CODE REVIEW

**Comment by rxin** [comment]
This is based on https://github.com/apache/spark/pull/7174

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  /**
+   * Enables matching against NumericType for expressions:
+   * {{{
+   *   case Cast(child @ NumericType(), StringType) =>
+   *     ...
+   * }}}
+   */
+  def unapply(e: Expression): Boolean = e.dataType.isInstanceOf[NumericType]
+
+  override def defaultConcreteType: DataType = IntegerType
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala]: Most of the rule seems like to be:
DecimalType => DecimalType
FractionalType => DoubleType
LongType => LongType
IntegralType => IntegerType

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 2]
```diff
@@ -704,19 +704,46 @@ object HiveTypeCoercion {
...
+
+        // Implicit cast among numeric types
+        case (in @ NumericType(), target: NumericType) if in.dataType != target =>
+          Cast(in, target)
+
+        // Implicit cast between date time types
+        case (in @ DateType(), TimestampType) => Cast(in, TimestampType)
+        case (in @ TimestampType(), DateType) => Cast(in, DateType)
+
+        // Implicit from string to atomic types, and vice versa
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: Seems this comment is not very consistent with the code below? For `case (in, StringType) if in.dataType != StringType`, we also cast complex types to string?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: good catch

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercionSuite.scala [hunk 3]
```diff
@@ -26,6 +26,31 @@ import org.apache.spark.sql.types._
 
 class HiveTypeCoercionSuite extends PlanTest {
 
+  test("implicit type cast") {
+    def shouldCast(from: DataType, to: AbstractDataType): Unit = {
+      val got = HiveTypeCoercion.ImplicitTypeCasts.implicitCast(Literal.create(null, from), to)
+      assert(got.dataType === to.defaultConcreteType)
+    }
+
+    // TODO: write the entire implicit cast table out for test cases.
```
Comment (yhuai) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercionSuite.scala]: Open a jira for this task?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  /**
+   * Enables matching against FractionalType for expressions:
+   * {{{
+   *   case Cast(child @ FractionalType(), StringType) =>
+   *     ...
+   * }}}
+   */
+  def unapply(e: Expression): Boolean = e.dataType.isInstanceOf[FractionalType]
+
+  override def defaultConcreteType: DataType = DoubleType
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala]: When will we use this method?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  /**
+   * Enables matching against IntegralType for expressions:
+   * {{{
+   *   case Cast(child @ IntegralType(), StringType) =>
+   *     ...
+   * }}}
+   */
+  def unapply(e: Expression): Boolean = e.dataType.isInstanceOf[IntegralType]
+
+  override def defaultConcreteType: DataType = IntegerType
```
Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala]: Should we use LongType as the default one?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 6]
```diff
@@ -174,8 +171,7 @@ case class Sha1(child: Expression) extends UnaryExpression with AutoCastInputTyp
  * A function that computes a cyclic redundancy check value and returns it as a bigint
  * For input of type [[BinaryType]]
  */
-case class Crc32(child: Expression)
-  extends UnaryExpression with AutoCastInputTypes {
+case class Crc32(child: Expression) extends UnaryExpression with ExpectsInputTypes {
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: Crc32 should be able to work with StringType, but StringType cannot be implicit casted BinaryType, right ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 7]
```diff
@@ -704,19 +704,46 @@ object HiveTypeCoercion {
...
+        case (in @ StringType(), target: AtomicType) if target != StringType =>
+          Cast(in, target.defaultConcreteType)
+        case (in, StringType) if in.dataType != StringType =>
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: Agree with @yhuai, we should consider more of the semantic context when cast to `StringType`.
e.g. `trim(value)`, the `value` should be exact the `StringType`.
   but `repeat(value, times)`, the `value` should be `StringType` or even any of the `NumericType`.

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: I don't think that makes sense w.r.t. "semantic context".

As a matter of fact, most mature databases have well defined semantics for implicit type casting. If we make numeric type -> stringtype an implicit cast rule, we should have that for all expressions. If we don't want that implicit cast rule, then we should never do the implicit cast. I'd argue that having some expressions doing implicit casting and some not doing it is extremely confusing.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala (hunk 6 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: i need to think about whether we should support implicit casts from string to binary. sql server does support that. hive doesn't, but hive chose to make a lot of the udfs work against both types.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala (hunk 7 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: btw "never" to a certain extent. I think there can be exceptions, but very few.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala (hunk 6 above)
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: I am not sure if we can always cast a string to binary correctly, as it produces different binary when specifying different encoder. It's actually the case accept multiple `DataType` for an expression. 
And also for `Length`, which support both `StringType` and `BinaryType`.

We probably need another PR for this improvement.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala (hunk 7 above)
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: Yes, that's what I was thinking, if we need to work with `Expression.checkInputTypes` or combine the logic.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala (hunk 6 above)
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: I was thinking about having an AbstractDataType that's a TypeCollection, that expressions can put arbitrary types into it. Basically similar to the Seq[Any] idea, but with better type safety.

Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: I think that's a good idea for this, but it probably make thing more complicated for auto casting. (Which data type should be cast to?)

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: +1 for StringType -> BinaryType

Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: I mean we'd better leave the casting (StringType -> BinaryType) to be done within the UDF `Crc32` itself, not via the generic auto casting rule. From the user perspective, the UDF Crc32 will support both `StringType` and `BinaryType`.

Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: Oh, sorry, I just checked the code of Hive, it does convert the StringType => BinaryType (UTF8 bytes), just as the generic rule. @davies  +1

**Comment by rxin** [comment]
I added an implicit type cast from String to Binary.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala [hunk 8]
```diff
@@ -704,19 +704,48 @@ object HiveTypeCoercion {
...
+        case (DateType, TimestampType) => Cast(e, TimestampType)
+        case (TimestampType, DateType) => Cast(e, DateType)
+
+        // Implicit cast from/to string
+        case (StringType, NumericType) => Cast(e, DoubleType)
+        case (StringType, target: NumericType) => Cast(e, target)
+        case (StringType, DateType) => Cast(e, DateType)
+        case (StringType, TimestampType) => Cast(e, TimestampType)
+        case (StringType, BinaryType) => Cast(e, BinaryType)
+        case (any, StringType) if any != StringType => Cast(e, StringType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/HiveTypeCoercion.scala]: according to hive https://cwiki.apache.org/confluence/display/Hive/LanguageManual+Types and discussion in https://github.com/apache/spark/pull/6551,
we should only allow atomic type(except boolean and binary) to string.

**Comment by cloud-fan** [comment]
lgtm

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+  // longer an no argument constructor and thus the JVM cannot serialize the object anymore.
+  private[sql] val numeric: Numeric[InternalType]
+}
+
+
+private[sql] object NumericType extends AbstractDataType {
+  /**
+   * Enables matching against NumericType for expressions:
+   * {{{
+   *   case Cast(child @ NumericType(), StringType) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala]: Exist: why we add unapply for it? Is it same with `Cast(child: NumericType, StringType)`? It looks to me that we only need this `object NumericType` in `ExpectsInputTypes` when an expression need any kind of numeric input.
And, should we add `object AtomicType` too?

Comment (yhuai) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala]: `child` is an expression.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala]: sorry didn't see that...

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/AbstractDataType.scala]: btw this is old code. just got copied around.

**Comment by rxin** [comment]
@marmbrus i'm going to merge this one since it blocks @davies' pr. I will submit a follow up pr to support type collections for inputs. Please continue to review and I will address feedback together.

**Pull request closed without merging**